WordPress/gutenberg
# Fix multi-entity multi-property undo redo

closes #12075

It's been sometime I didn't dig into the undo/redo reducer and while taking a deep look into it, I noticed a fundamental issue in how the stack is recorded and applied when undoing/redoing.

### How Undo works ?

To understand the issue and the need for this PR to solve it, we first need to understand how the undo/redo works in Gutenberg. 

In trunk, for any change done in the editor, a "set of edits" is added to a list of undos. Each edit, contain the following information: The identifier of the modified entity/record + the set of edits to apply when going (either back or forth) to that state. So the list resembles something like this:

 - entity1, { property1: value1, property2: value2 }
 - entityX { propertyX: valueX, propertyY: valueY }
 - ...

We also keep track of a "pointer" to a particular "step" in this list. By default the pointer is always pointing at the last step (current step) and when calling "undo" it goes back by 1 step (and opposite for redo). 

Each time this step is modified (undo or redo called), we retrieve the set of "edits" corresponding to that step and apply it.

### How do these steps gets recorded

This is a bit complex (as there are edge cases) but for the regular case, the behavior is the following: Let's stay initially the "undo list" contained the following content:

- entity1, { property1: value1 }

The user in the UI edits property1 with value12, the list becomes:

- entity1, { property1: value1 }
- entity1, { property1: value12 }

The user in the UI edits property2 with value22 (let's also assume that the previous value of value2 was value21 but this was not in the undo list, it's just the saved value), the list becomes

- entity1, { property1: value1 }
- entity1, { property2: value21 }
- entity1, { property2: value22 }

Notice that for the second item (penultimate item) of the list, we actually updated the edits to store the previous value of  property2. The reason we do this is because when "undoing" to that step, we need to restore the previous value, so we need to "edit" value2 to set the previous value.

And here's the bug, now let's say with used undo twice, we went to the initial step (property 1 is value1), and now we "redo" to move to the pointer to step2. What will happen now is that the edits of step2 will be applied, so `property2 set to value21` but if you remember properly, in step2, property1 need to be set to value12 and it didn't happen. (This is the bug recorded in #12075)

So what this tells us is that for each "step" in the undo list, we need to record both:

 - The edits to apply when we go up the list (redo)
 - The edits to apply when we go down the list (undo)

### Multi-entity bug

Worth nothing that in trunk, the undo/redo suffer also from the "multi-entity" diff problem. Each diff can contain multiple properties (especially when using transient edits) but in theory "multi-entity" diffs are possible. So each "diff" should be a list of diffs instead.

### Solution

One solution to this is to update the "diff" format to use a (from, to) tupples instead, so for the example above, we should have the following undo stack:

- [ { entity1,  property1, from: value1, to: value12 } ]
- [ { entity2,  property2, from: value21, to: value22 } ]

And when undoing, we use the pointer to retrieve the current diff and edit the entities using the "from" values. (Restore the from values), while for redos, we'd use the "to" values.

Now, this solves navigating up and down the list and we'll get the expected result (and works both for multi property and multi entities).

**This is the main change in this PR** 

 - It changes the format of the undo list in the reducer
 - It allows "multiple edits from multiple entities" to be applied at the same time (using UNDO, REDO actions for now)
 - It also contains secondary changes to simplify the reducer: more specifically "flattenedUndo" is replaced with "cache" which represents a list of multi entity edits that have been performed but no "undo" step has been created for them, they are just waiting for the "next" undo step to be created. This is needed to support "transient entity changes" and "stacked undo steps in RichText".
 - Unfortunately, since the format of the list change and the format is currently exposed in `getUndoEdit`/`getRedoEdit`, this PR contains a breaking change right now. Ideally these selectors should have been private, I'll find a way to solve this by using new private selectors and trying to return something similar in format for these two selectors and probably deprecate them.

There are some extra simplifications that are possible to our undo/redo APIs but are not included in this PR.

**Todo**

 - [x] Ensure all existing unit/e2e tests pass.
 - [x] Add a unit test and/or e2e test for the specific issue being fixed.
 - [x] Backward compatibility for getUndoEdit / getRedoEdit

**Testing instructions**

 - Create a new post
 - Type a character in the post title
 - Remove that character from the post title
 - Type something in the post content
 - Undo until the post title is filled back with the initial character
 - Redo ==> the character should get removed again.
